core/shared/src/main/scala/zio/prelude/ZValidation.scala

```diff
@@ -75,8 +75,8 @@ sealed trait ZValidation[+W, +E, +A] { self =>
    */
   final def flatMap[W1 >: W, E1 >: E, B](f: A => ZValidation[W1, E1, B]): ZValidation[W1, E1, B] =
     self match {
-      case Failure(w, e) => Failure(w, e)
-      case Success(w, a) =>
+      case failure @ Failure(_, _) => failure
```

Comment:
From a performance perspective, I think this is probably better as it avoids extracting the case class params:

```scala
case failure: Failure[E1, B] => failure
```
Depending on the scalac flags this might raise a warning (I think if the `unchecked` flag is used) , in which case you can do this:

```scala
case failure: Failure[E1 @unchecked, B @unchecked] => failure

// Or alias the annotation to reduce boilerplate:
import scala.{ unchecked => uc }
```

Reply:
Indeed this.